### ISSUE TITLE
Support namespaces for types and values

### ISSUE THREAD

**Issue by matklad** [intellij-rust/intellij-rust#552]
Rust has separate namespaces for types and values. Because of this, the following is perfectly valid:

```Rust
pub struct P<T: ?Sized> { ptr: Box<T> }

#[allow(non_snake_case)]
pub fn P<T: 'static>(value: T) -> P<T> {
    P { ptr: Box::new(value) }
}
```

One `P` is a struct **type**, another one is fn **value**, so there is no collision. 

Currently we use a single namespace, and it works well in practice, because such ^^^^ shadowing is extremely rare in practice. It would be nice to support it though!

**Comment by matklad** [comment]
https://github.com/nrc/rfcs/blob/6098278299d00d184d5b735ca6751c3e1765b406/text/0000-name-resolution.md

**Comment by Furyhunter** [comment]
The following is a case where that should cause an error inspection in IDEA, the fn is considered defined first here and the compiler throws an error.

```rust
pub struct P(pub i32);

pub fn P(value: i32) -> P { P(value) }
```

**Comment by matklad** [comment]
Yep, declaring tuple struct actually declares a tuple *type* and a function with the same name, which construct the structs. This is similar to unit struct type/const value.

**Comment by kumbayo** [comment]
In my code example a module path(with the same name as a binding) is erroneously highlighted.
is this caused by the same issue or should i create a new Issue for this?

```rust
mod test {
    pub struct Test {
        pub a: u32,
    }
}

fn main() {
    let mut test = test::Test { a: 42 };
    // <- Here we would mutably use test
    let test: test::Test = test; // New immutable binding so test is not accidentally modified
    println!("Value: {}", test.a);
}
```

![intellij_module_path_highlighted_as_binding](https://cloud.githubusercontent.com/assets/1041267/17524520/49e4c002-5e5f-11e6-9c7d-7f6cdfa23607.png)

**Comment by matklad** [comment]
Yes, this is the same issue. Thanks for the not so esoteric example!

**Comment by matklad** [comment]
Basic support for namespaces has landed in https://github.com/intellij-rust/intellij-rust/commit/f5d12ee2875c49e1fccb4b0e51587469cb0aef3d
